crates/services/p2p/src/codecs/gossipsub.rs

```diff
@@ -0,0 +1,53 @@
+use std::io;
+
+use fuel_core_types::fuel_tx::Transaction;
+
+use crate::gossipsub::messages::{
+    GossipTopicTag,
+    GossipsubBroadcastRequest,
+    GossipsubMessage,
+};
+
+use super::{
+    Decode,
+    Encode,
+    Encoder,
+    GossipsubCodec,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct GossipsubMessageHandler<Codec> {
+    pub(crate) codec: Codec,
+}
+
+impl<Codec> GossipsubCodec for GossipsubMessageHandler<Codec>
+where
+    Codec: Encode<Transaction, Error = io::Error>
+        + Decode<Transaction, Error = io::Error>
+        + Send,
+{
+    type RequestMessage = GossipsubBroadcastRequest;
+    type ResponseMessage = GossipsubMessage;
+
+    fn encode(&self, data: Self::RequestMessage) -> Result<Vec<u8>, io::Error> {
+        match data {
+            GossipsubBroadcastRequest::NewTx(tx) => {
+                Ok(self.codec.encode(&tx)?.as_bytes().into_owned())
```

Comment:
If you convert it into owned, I think it will be better to return vector from the `Encoder` trait itself. Because right now you create two vectors.

Reply:
I prefer the Cow over the Vec for these low-level traits. A Cow can be converted to a vec without cloning using `Vec::from`. See https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-From%3CCow%3C'a,+%5BT%5D%3E%3E-for-Vec%3CT%3E

So the problem here is the `.as_bytes().into_owned()` rather than the use `Cow`.